i#2297: AARCH64: Implement mbr & cbr instrumentation

jiegec opened the pull request:
Implement mbr & cbr instrumentation for AARCH64, previously only available for X86. For cbr instrumentation, there are three types of conditional branches on AARCH64:

1. conditional branch if (non) zero: generate code to compare the register against zero to compute the direction
2. test bit and branch if (non) zero: generate code to extract the bit out of the register (using ubfm) and then compare the bit against zero to compute the direction
3. conditional branch according to NZCV: generate code to compute the direction using csinc instruction

Special care is implemented if the stolen register x28 is used as the operand, to avoid generating incorrect results.

For mbr instrumentation, the target address is placed in the register, which is later passed to the clean call from the user.

Additional instruction creation macros are added for ubfm and csinc instructions. Support for immediate operand is added to the INSTR_CREATE_eor macro. XINST_CREATE_slr_s is updated to use the new INSTR_CREATE_ubfm macro.

Count-ctis test is now enabled due to cbr & mbr instrumentation available on AARCH64. It has been enhanced to validate the direction and target address of conditional branches to validate the instrumentation.

The instrcalls and cbrtrace example clients are enabled as well. They are tested on real AARCH64 hardware. Additionally, a manually crafted dynamorio client can successfully capture the whole control flow transfer history with this pull reuqest.

Fixes: #2297 (only for AARCH64, ARM not yet) and #2919.

AssadHashmi commented on core/ir/aarch64/instr_create_api.h:
```diff
@@ -658,13 +658,19 @@
-#define INSTR_CREATE_eor(dc, d, s)                                      \
-    INSTR_CREATE_eor_shift(dc, d, d, s, OPND_CREATE_INT8(DR_SHIFT_LSL), \
-                           OPND_CREATE_INT8(0))
+#define INSTR_CREATE_eor(dc, d, s_or_imm)                                            \
```
Thank you for contributing.

Ideally, every new addition of an `INSTR_CREATE_` or `XINSTR_CREATE_` should include the Doxygen comment block describing the macro, e.g.
```
/**                                                                       
 * Creates an EOR instruction with one output and two inputs.             
 * \param dc   The void * dcontext used to allocate memory for the instr_t.
 * \param rd   The output register.                                       
 * \param rn   The first input register.                                  
 * \param rm_or_imm   The second input register or immediate.             
 */
```

jiegec replied:
Thanks, added

AssadHashmi commented on core/ir/aarch64/instr_create_api.h:
```diff
@@ -658,13 +658,19 @@
                            opnd_add_flags(sht, DR_OPND_IS_SHIFT), sha)
+#define INSTR_CREATE_csinc(dc, rd, rn, rm, cond) \
+    instr_create_1dst_3src(dc, OP_csinc, rd, rn, rm, cond)
+#define INSTR_CREATE_ubfm(dc, rd, rn, lsb, width) \
```
Please replace the `instr_create_1dst_3src(dc, OP_ubfm...` in `XINST_CREATE_slr_s()` with this `INSTR_CREATE_ubfm()`.

jiegec replied:
Thanks, done

derekbruening commented on core/ir/aarch64/instr_create_api.h:
```diff
@@ -463,12 +463,12 @@
  * they just need to know whether they need to preserve the app's flags, so maybe
```
Please enable the samples that use this which are currently disabled:
```
  # dr_insert_cbr_instrument_ex is NYI
  add_sample_client(cbrtrace    "cbrtrace.c;utils.c"    "drmgr;drx")
  add_sample_client(hot_bbcount "hot_bbcount.c"         "drmgr;drreg;drbbdup;drx")
```

Those are run as tests, though w/o targeted correctness checks: just making sure they don't crash.

derekbruening replied:
Please update the PR description to describe how you tested this, since there are no tests added/enabled by this PR (see prior comment which will add some).

derekbruening commented on core/ir/aarch64/instr_create_api.h:
```diff
@@ -463,12 +463,12 @@
  * they just need to know whether they need to preserve the app's flags, so maybe
```
Please enable the count-ctis tests which use this, which will add regression tests.

derekbruening replied:
Oh, didn't realize it needed more: was just grepping for tests that use cbr.  Makes sense to separate out the mbr.  Is it easy to separate in the test?  Or just separately locally in the test and confirm cbr works and state that in the PR description and say that the test will be enabled soon when mbr is added and then enable the test in a separate PR for mbr, so long as that comes in relatively soon (i.e., not months later with no cbr test in the meantime).

mbr is supposed to obtain a real address so that sounds like something is wrong there.

jiegec replied:
I have found the problem: the register storing the actual target address is clobbered when saving registers to mcontext, so when it is passed to the clean call, the register content is invalid. How can I avoid the register being clobbered? For x86, it seems that the TLS operand is used instead as argument to the clean call, but it does not work on ARM64.

jiegec replied:
Reading `dr_insert_cbr_instrumentation_help`, it seems the the solution is to insert instructions between saving registers and calling clean call, which is hard to implement, hmmm.

derekbruening replied:
> I have found the problem: the register storing the actual target address is clobbered when saving registers to mcontext, so when it is passed to the clean call, the register content is invalid.

Sorry, could you elaborate to help remind me of what is going on: who is saving registers to mcontext?  Is that the clean call setup?  It shouldn't be: if you pass your register to the clean call code it should arrange to preserve it.  It must be something else saving to mcontext?

jiegec replied:
> > I have found the problem: the register storing the actual target address is clobbered when saving registers to mcontext, so when it is passed to the clean call, the register content is invalid.
> 
> Sorry, could you elaborate to help remind me of what is going on: who is saving registers to mcontext? Is that the clean call setup? It shouldn't be: if you pass your register to the clean call code it should arrange to preserve it. It must be something else saving to mcontext?

Yes, the clean call setup code is saving registers to mcontext (the call to 0x4118c580). Otherwise, there are no instructions that might clobber the registers.

derekbruening replied:
That sounds like a bug in the AArch64 clean call code: if the user specifies a register as a clean call arg, the clean call setup has to preserve that register and pass it to the callee.

I think x86 does an xchg to TLS not because the clean call code will mess with the register but to get the register state to have app values for the callee using dr_get_mcontext.  For most clients: they should use drreg and this will be done automatically by DR_CLEANCALL_READS_APP_CONTEXT which is set for this mbr routine.

All that said: now I'm wondering whether drreg restoring for DR_CLEANCALL_READS_APP_CONTEXT would clobber a param value -- don't remember enough details there: maybe @abhinav92003 knows; or we can study the code or do an experiment.

Plus, here you're not using drreg as you're inside the core: so this seems like a bug.  It may be worth reviewing the x86 clean call setup code and confirm it avoids clobbering and see how it does it, and compare to AArch64.  If confirmed that this is a general AArch64 clean call bug: please file a separate issue on it as it could affect other users.

jiegec replied:
> That sounds like a bug in the AArch64 clean call code: if the user specifies a register as a clean call arg, the clean call setup has to preserve that register and pass it to the callee.
> 
> I think x86 does an xchg to TLS not because the clean call code will mess with the register but to get the register state to have app values for the callee using dr_get_mcontext. For most clients: they should use drreg and this will be done automatically by DR_CLEANCALL_READS_APP_CONTEXT which is set for this mbr routine.
> 
> All that said: now I'm wondering whether drreg restoring for DR_CLEANCALL_READS_APP_CONTEXT would clobber a param value -- don't remember enough details there: maybe @abhinav92003 knows; or we can study the code or do an experiment.
> 
> Plus, here you're not using drreg as you're inside the core: so this seems like a bug. It may be worth reviewing the x86 clean call setup code and confirm it avoids clobbering and see how it does it, and compare to AArch64. If confirmed that this is a general AArch64 clean call bug: please file a separate issue on it as it could affect other users.

Sorry, it is a bug in my testing code of dynamorio client for mbr instrumentation. The current code works correctly.

derekbruening commented on core/ir/aarch64/instr_create_api.h:
```diff
@@ -658,14 +658,49 @@
+    instr_create_1dst_3src(dc, OP_csinc, rd, rn, rm, cond)
+
+/**
+ * Creates an UBFM instruction with one output and three inputs.
```
(nit: Do you pronounce this "uhb-fmmm"?  My brain doesn't like the "an" because I say it "you-bee-eff-emm" and so I want "a" not "an" but if you start it with "uh" feel free to leave the "an".)

derekbruening commented on core/lib/instrument.c:
```diff
@@ -6328,6 +6328,146 @@ dr_insert_cbr_instrumentation_help(void *drcontext, instrlist_t *ilist, instr_t
+    reg_id_t dir = DR_REG_NULL;
+    reg_id_t flags = DR_REG_NULL;
+    int opc;
+    ;
```
Please remove.

derekbruening commented on core/lib/instrument.c:
```diff
@@ -6328,6 +6328,146 @@ dr_insert_cbr_instrumentation_help(void *drcontext, instrlist_t *ilist, instr_t
+                  "dr_insert_cbr_instrumentation must be applied to a cbr");
+    target = (ptr_uint_t)opnd_get_pc(instr_get_target(instr));
+
+    /* Compute branch direction */
```
style: For all new comments, please capitalize and use end-punctuation: so end period here, and capitalize and add end periods to all the ones below: line 6352, line 6354, etc.

derekbruening commented on core/lib/instrument.c:
```diff
@@ -6328,6 +6328,146 @@ dr_insert_cbr_instrumentation_help(void *drcontext, instrlist_t *ilist, instr_t
+        opnd_t reg_op = instr_get_src(instr, 1);
+        reg_id_t reg = opnd_get_reg(reg_op);
+        /* use dir register to compute direction */
+        dir = (reg == DR_REG_X0 || reg == DR_REG_W0) ? DR_REG_X1 : DR_REG_X0;
```
IMHO using `reg_to_pointer_sized(reg) == DR_REG_X0` is better than checking both X0 and W0 since it will generalize to other architectures or SIMD registers which have more than 2 overlapping.  Ditto below.

derekbruening commented on core/lib/instrument.c:
```diff
@@ -6328,6 +6328,146 @@ dr_insert_cbr_instrumentation_help(void *drcontext, instrlist_t *ilist, instr_t
+        flags = (reg == DR_REG_X2 || reg == DR_REG_W2) ? DR_REG_X3 : DR_REG_X2;
+        /* save old value of flags register to SPILL_SLOT_2 */
+        dr_save_reg(dcontext, ilist, instr, flags, SPILL_SLOT_2);
+        /* save flags to flags register */
```
Hmm, instead of using CMP and having to save the flags is it better to branch and use the same CBZ/CBNZ to set dir_op to 0 or 1?  Fewer loads+stores, but has a branch...my intuition says avoiding the store is worth having the branch.  But probably should be profiled.  OK to just put a XXX comment about it.

derekbruening commented on core/lib/instrument.c:
```diff
@@ -6328,6 +6328,146 @@ dr_insert_cbr_instrumentation_help(void *drcontext, instrlist_t *ilist, instr_t
+        reg_id_t dir_same_width = DR_REG_NULL;
+
+        /* use dir register to compute direction */
+        if (DR_REG_START_64 <= reg && reg <= DR_REG_STOP_64) {
```
Use reg_is_64bit()

derekbruening commented on core/lib/instrument.c:
```diff
@@ -6328,6 +6328,146 @@ dr_insert_cbr_instrumentation_help(void *drcontext, instrlist_t *ilist, instr_t
+        } else {
+            /* 32-bit register */
+            dir = (reg == DR_REG_W0) ? DR_REG_X1 : DR_REG_X0;
+            dir_same_width = (reg == DR_REG_W0) ? DR_REG_W1 : DR_REG_W0;
```
Looks like you can use API routines to make this simpler, replacing this if/else lines 6378-6386 with this:
```
dir = (reg_to_pointer_sized(reg) == DR_REG_X0) ? DR_REG_X1 : DR_REG_X0;
dir_same_width = reg_resize_to_opsz(dir, opnd_get_size(reg));
```

derekbruening reviewed (commented):
Thank you for the contribution.  The main missing piece is testing: see the comments.

jiegec commented:
Help is wanted to fix mbr instrumentation: it sometimes receives small integers instead of actual target address, possibly some indices into the basic box cache?

derekbruening commented:
> Help is wanted to fix mbr instrumentation: it sometimes receives small integers instead of actual target address, possibly some indices into the basic box cache?

I assume it has the wrong register or something.  There are no such "indices" in in-fragment code.  Use the logs https://dynamorio.org/page_logging.html to see where the app's register values goes and which one is passed to the clean call.

jiegec commented:
Newly enabled count-ctis test is passed. Manually written dynamorio client to leverage cbr/mbr instrumentation also works correctly.

jiegec commented:
A bug has been found: when `x28` is used for `cbnz` conditional branch, the branch direction is wrongly computed. It is possibly due to x28 being used by dynamorio.

jiegec commented:
> A bug has been found: when `x28` is used for `cbnz` conditional branch, the branch direction is wrongly computed. It is possibly due to x28 being used by dynamorio. [dr_insert_get_stolen_reg_value()](https://dynamorio.org/dr__ir__utils_8h.html#a7a38641e89e2573d725ec05d137e0fb6) is required to fix this.

Fixed in commit e9980ab.

derekbruening commented:
Please click the re-request review in the upper right when ready (comments resolved and tests green).

derekbruening commented:
> @[jiegec](https://github.com/jiegec) jiegec [force-pushed](https://github.com/DynamoRIO/dynamorio/compare/f02728e2cc48a04663d339fb180c41d8282ff258..d9222e144c1c6c2d092358e2763e338519770e03) the i2297-cbr-aarch64 branch from [f02728e](https://github.com/DynamoRIO/dynamorio/commit/f02728e2cc48a04663d339fb180c41d8282ff258) to [d9222e1](https://github.com/DynamoRIO/dynamorio/commit/d9222e144c1c6c2d092358e2763e338519770e03) 
[3 days ago](https://github.com/DynamoRIO/dynamorio/pull/7005#event-17601062475)

Please do not force-push shared branches. This destroys the history and makes it a lot harder for reviewers!  This is documented in several places: https://dynamorio.org/page_contributing.html, https://dynamorio.org/page_code_reviews.html#autotoc_md118.  There is now no way to easily see the changes since the last round of reviews.  You're forcing the reviewer to start from scratch; throwing away the anchors to all the prior comments; etc.

jiegec commented:
> > @[jiegec](https://github.com/jiegec) jiegec [force-pushed](https://github.com/DynamoRIO/dynamorio/compare/f02728e2cc48a04663d339fb180c41d8282ff258..d9222e144c1c6c2d092358e2763e338519770e03) the i2297-cbr-aarch64 branch from [f02728e](https://github.com/DynamoRIO/dynamorio/commit/f02728e2cc48a04663d339fb180c41d8282ff258) to [d9222e1](https://github.com/DynamoRIO/dynamorio/commit/d9222e144c1c6c2d092358e2763e338519770e03)
> > [3 days ago](https://github.com/DynamoRIO/dynamorio/pull/7005#event-17601062475)
> 
> Please do not force-push shared branches. This destroys the history and makes it a lot harder for reviewers! This is documented in several places: https://dynamorio.org/page_contributing.html, https://dynamorio.org/page_code_reviews.html#autotoc_md118. There is now no way to easily see the changes since the last round of reviews. You're forcing the reviewer to start from scratch; throwing away the anchors to all the prior comments; etc.

I'm sorry for that. It spans too long time for this pull request for me to remember the rules..

derekbruening commented on core/ir/aarch64/instr_create_api.h:
```diff
@@ -658,14 +658,49 @@
+ * \param d         The output register and the first input register.
+ * \param s_or_imm  The second input register or immediate.
+ */
+#define INSTR_CREATE_eor(dc, d, s_or_imm)                                            \
```
Please list these 2 added macros and 2 updated macros in the PR description == final commit message.

jiegec replied:
Done.

derekbruening commented on core/lib/instrument.c:
```diff
@@ -6129,6 +6129,32 @@ dr_insert_mbr_instrumentation(void *drcontext, instrlist_t *ilist, instr_t *inst
+
+    /* For all known mbr instructions on aarch64, the first source register saves the
+     * target. */
+    target = opnd_get_reg(instr_get_src(instr, 0));
```
Use instr_get_target()

jiegec replied:
Done.

derekbruening commented on suite/tests/client-interface/count-ctis.c:
```diff
@@ -95,5 +96,35 @@ GLOBAL_LABEL(FUNCNAME:)
+        mov      x1, ARG1
+        END_PROLOG
+
+        /* test cbnz */
```
Comment says cbnz but I only see cbz.

jiegec replied:
Added.

derekbruening commented on suite/tests/client-interface/count-ctis.c:
```diff
@@ -95,5 +96,35 @@ GLOBAL_LABEL(FUNCNAME:)
+        mov      x1, 1
+        cbz      x1, jecxz_taken
+        nop
+     jecxz_not_taken:
```
Please add a test case here for each condition in your code:

+ cbz using stolen register
+ cbnz
+ cbnz using stolen register
+ tbz
+ tbz using stolen register
+ tbnz
+ bcond

(Even without a comprehensive register state check before vs after each, which is not simple to add, just having these exercise each part of the instrumentation code is a good step.  And errors in this type of instrumentation often result in visible crashes.  Could add some aarch64-specific testing in the client for further tests but that is optional.  Just adding app code for each of the cases listed should be simple and go a long way.)

jiegec replied:
Done.

derekbruening reviewed (approved):
Looks good if additional sanity test cases are added which should be straightforward.

jiegec commented on suite/tests/client-interface/count-ctis.c:
```diff
@@ -105,15 +105,112 @@ GLOBAL_LABEL(FUNCNAME:)
+        cmp      x1, 1
+        bne      bne_taken
+        nop
+     bne_taken:
```
Done:

To test the new cbr instrumentation of AARCH64, add additional branch pc
check: dst == src + 0x8, and direction check to count-ctis. Region of
interest is marked via nop; nop; add; nop sequence. The triple nops
pattern does not work, because it also appears in other places.

derekbruening reviewed (approved):
LGTM: Thank you for improving the test, this gives more confidence the API works and gives us a good regression test to detect unforeseen issues with future code changes.

derekbruening commented:
I've sent you an invite to gain commit access.  You can then merge here (need to resolve all the conversations by clicking Resolve first).  You can then use branches on this DynamoRIO/dynamorio repository directly rather than a clone.
